controller/volume_controller_test.go

```diff
@@ -820,6 +821,72 @@ func (s *TestSuite) TestVolumeLifeCycle(c *C) {
 	}
 	tc.expectReplicas = expectRs
 	testCases["volume attaching - start replicas - node failed"] = tc
+
+	// Disable revision counter
+	tc = generateVolumeTestCaseTemplate()
+	tc.volume.Spec.NodeID = TestNode1
+	tc.volume.Status.OwnerID = TestNode1
+	tc.volume.Status.CurrentNodeID = TestNode1
+
+	// Since default node 2 is scheduling disabled,
+	// enable node soft anti-affinity for 2 replicas.
+	tc.replicaNodeSoftAntiAffinity = "true"
+	tc.volume.Spec.RevisionCounterDisabled = true
+	tc.copyCurrentToExpect()
+	tc.engines = nil
+	tc.replicas = nil
+
+	tc.expectVolume.Status.State = types.VolumeStateAttaching
+	tc.expectVolume.Status.CurrentImage = tc.volume.Spec.EngineImage
+	tc.expectVolume.Status.CurrentNodeID = tc.volume.Spec.NodeID
+	expectEs := map[string]*longhorn.Engine{}
+	for _, e := range tc.expectEngines {
+		e.Spec.RevisionCounterDisabled = true
+		expectEs[e.Name] = e
+	}
+	tc.expectEngines = expectEs
+	expectRs = map[string]*longhorn.Replica{}
+	for _, r := range tc.expectReplicas {
+		r.Spec.DesireState = types.InstanceStateRunning
+		r.Spec.RevisionCounterDisabled = true
+		expectRs[r.Name] = r
+	}
+	tc.expectReplicas = expectRs
+
+	testCases["volume revision counter disabled - engine and replica revision counter disabled"] = tc
+
+	// Salvage Requested
+	tc = generateVolumeTestCaseTemplate()
+	tc.volume.Spec.NodeID = TestNode1
+	tc.volume.Status.State = types.VolumeStateDetached
+	tc.volumeAutoSalvage = "true"
+
+	for _, e := range tc.engines {
+		e.Status.CurrentState = types.InstanceStateStopped
+	}
+	for _, r := range tc.replicas {
+		r.Spec.FailedAt = getTestNow()
```

Comment:
The init state should be:
```
r.Status.CurrentState = types.InstanceStateError
r.Spec.healthyAt = getTestNow()
......
```

Then the desired state should be:
```
r.Spec.FailedAt = getTestNow()
r.Spec.DesiredState = types.InstanceStateStopped
......
```

Reply:
I think to trigger the auto-salvage, all replicas need to be failed from [this code](https://github.com/longhorn/longhorn-manager/blob/38c9c683ce63bf29afac0b0c8649626a6102dae0/controller/volume_controller.go#L869-L873). This is why I set the init state with 'FailedAt'.

Reply:
The prerequisite of the auto salvage is that all replicas become error. Considering the following workflow:
1. All replica processes become `error` simultaneously.
2. The instance handler in Replica Controller will update the `replica.status.state` to `error` accordingly. (Here, `replica.Spec.FailedAt` is still empty. And I think this should be the init state of this unit test.)
3. The Volume Controller get triggered. In the **single** sync loop:
    1. In `ReconcileEngineReplicaState`, `replica.Spec.FailedAt` will be set.
    2. In `ReconcileVolumeState`, the replicas will be salvaged immediately. Both `v.Status.CurrentNodeID` and `replica.Spec.FailedAt` [will be unset](https://github.com/longhorn/longhorn-manager/blob/38c9c683ce63bf29afac0b0c8649626a6102dae0/controller/volume_controller.go#L910).
    3. `e.Spec.DesireState = types.InstanceStateStopped` 
    ......

The state I provided above maybe not completely precise/correct. But you cannot use the transit state in the sync loop as the init state of a unit test

Reply:
Hi Shuo, thanks for explaining this in details. And it totally make sense. I have updated the PR, please take a look. Thanks.